src/Geometry/RefinedDiscreteModels.jl

```diff
@@ -0,0 +1,298 @@
+  )
+  cell_node_ids_ccw = vcat(cell_node_ids'...)
+  #cell_node_ids_ccw = cell_node_ids
+  #@show cell_node_ids_ccw
+  for (i, cell) in enumerate(cell_node_ids)
+    cell_coords = node_coords[cell]
+    perm = sort_ccw(cell_coords)
+    cell_node_ids_ccw[i, :] = cell[perm]
+  end
+  cell_node_ids_ccw
+end
+
+"""
+node_coords == node, cell_node_ids == elem in Long Chen's notation
+"""
+function newest_vertex_bisection(
+  top::GridTopology,
+  node_coords::AbstractArray{<:VectorValue},
+  cell_node_ids::Matrix{Ti},
+  η_arr::AbstractArray{<:AbstractFloat},
+  θ::AbstractFloat,
+  sort_flag::Bool,
+) where {Ti <: Integer}
+  # Number of nodes
+  N::Ti = size(node_coords, 1)
+  elem = cell_node_ids
+  # Number of cells (triangles in 2D)
+  NT::Ti = size(elem, 1)
+  @assert length(η_arr) == NT
+  if sort_flag
+    sort_longest_edge!(elem, node_coords, NT)
+  end
+  # Make sure elem is consistent with GridTopology
+  test_against_top(elem, top, 2)
+  edge = build_edges(elem)
+  NE::Ti = size(edge, 1)
+  dualedge = build_directed_dualedge(elem, N, NT)
+  d2p = dual_to_primal(edge, NE, N)
+  # Make sure edge is consistent with GridTopology
+  test_against_top(edge, top, 1)
+  node_coords, marker =
+    setup_markers_and_nodes!(node_coords, elem, d2p, dualedge, NE, η_arr, θ)
+  cell_node_ids = bisect(d2p, elem, marker, NT)
+  node_coords, cell_node_ids
+end
+
+# step 1
+function newest_vertex_bisection(
+  grid::Grid,
+  top::GridTopology,
+  η_arr::AbstractArray{<:AbstractFloat},
+  θ::AbstractFloat,
+  sort_flag::Bool,
+)
+  node_coords = get_node_coordinates(grid)
+  cell_node_ids = get_cell_node_ids(grid)
+  if sort_flag
+    cell_node_ids_ccw = sort_cell_node_ids_ccw(cell_node_ids, node_coords)
+  else
+    cell_node_ids_ccw = vcat(cell_node_ids'...)
+  end
+  node_coords_ref, cell_node_ids_ref =
+    newest_vertex_bisection(top, node_coords, cell_node_ids_ccw, η_arr, θ, sort_flag)
+  # TODO: Should not convert to matrix and back to Table
+  cell_node_ids_ref = Table([c for c in eachrow(cell_node_ids_ref)])
+  reffes = get_reffes(grid)
+  cell_types = get_cell_type(grid)
+  # TODO : Gracefully handle cell_types
+  new_cell_types = fill(1, length(cell_node_ids_ref) - length(cell_node_ids))
+  cell_types = [cell_types; new_cell_types]
+  UnstructuredGrid(node_coords_ref, cell_node_ids_ref, reffes, cell_types)
+end
+
+# step 2
+function newest_vertex_bisection(
+  model::DiscreteModel,
+  η_arr::AbstractArray{<:AbstractFloat};
+  θ = 1.0, # corresponds to uniform refinement
+  sort_flag = false,
+)
+  # Not sure if necessary to keep old model unchanged. For my tests I use this
+  model_c = deepcopy(model)
+  grid = get_grid(model_c)
+  top = get_grid_topology(model_c)
+  ref_grid = newest_vertex_bisection(grid, top, η_arr, θ, sort_flag)
+  ref_topo = GridTopology(ref_grid)
+  #ref_labels = # Compute them from the original labels (This is perhaps the most tedious part)
+  ref_labels = FaceLabeling(ref_topo)
```

Comment:
As you mentioned, this forgets about the old labels. Ideally, one wants to refine the given labels instead of creating new unrelated ones.